proof.go

```diff
@@ -465,6 +465,212 @@ func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutN
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		// if the current range does not overlap with the proof range, pop and
+		// return a proof node if present, else return nil because subtree
+		// doesn't exist
+		if end <= proof.Start() || start >= proof.End() {
+			return popIfNonEmpty(&proof.nodes), nil
+		}
+
+		if len(ranges) != 0 && ranges[0].start == start && ranges[0].end == end {
+			ranges = ranges[1:]
+			return popIfNonEmpty(&subtreeRoots), nil
+		}
+
+		// Recursively get left and right subtree
+		k := getSplitPoint(end - start)
+		left, err := computeRoot(start, start+k)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start, start+k, err)
+		}
+		right, err := computeRoot(start+k, end)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start+k, end, err)
+		}
+
+		// only right leaf/subtree can be non-existent
+		if right == nil {
+			return left, nil
+		}
+		hash, err := nth.HashNode(left, right)
+		if err != nil {
+			return nil, fmt.Errorf("failed to hash node: %w", err)
+		}
+		return hash, nil
+	}
+
+	// estimate the leaf size of the subtree containing the proof range
+	proofRangeSubtreeEstimate := getSplitPoint(proof.End()) * 2
+	if proofRangeSubtreeEstimate < 1 {
+		proofRangeSubtreeEstimate = 1
+	}
+	rootHash, err := computeRoot(0, proofRangeSubtreeEstimate)
+	if err != nil {
+		return false, fmt.Errorf("failed to compute root [%d, %d): %w", 0, proofRangeSubtreeEstimate, err)
+	}
+	for i := 0; i < len(proof.Nodes()); i++ {
+		rootHash, err = nth.HashNode(rootHash, proof.Nodes()[i])
+		if err != nil {
+			return false, fmt.Errorf("failed to hash node: %w", err)
+		}
+	}
+
+	return bytes.Equal(rootHash, root), nil
+}
+
+// ToLeafRanges returns the leaf ranges corresponding to the provided subtree roots.
+// It uses the subtree root threshold to calculate the maximum number of leaves a subtree root can
+// commit to.
+// The subtree root threshold is defined as per ADR-013:
+// https://github.com/celestiaorg/celestia-app/blob/main/docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md
+// This method assumes:
+// - The subtree roots are created according to the ADR-013 non-interactive defaults rules
+// - The tree's number of leaves is a power of two
+// The algorithm is as follows:
+// - Let `d` be `y - x` (the range of the proof).
+// - `i` is the index of the next subtree root.
+// - While `d != 0`:
+//   - Let `z` be the largest power of 2 that fits in `d`; here we are finding the range for the next subtree root.
+//   - The range for the next subtree root is `[x, x + z)`, i.e., `S_i` is the subtree root of leaves at indices `[x, x + z)`.
+//   - `d = d - z` (move past the first subtree root and its range).
+//   - `i = i + 1`.
+//   - Go back to the loop condition.
+//
+// Note: This method is Celestia specific.
+func ToLeafRanges(proofStart, proofEnd, subtreeRootThreshold int) ([]LeafRange, error) {
```

Comment:
This is also exported so that we have a way downstream to find the leaf ranges based on a proof start -> end